api/provision/csp-write-portfolio.ts

```diff
@@ -6,87 +6,117 @@ import inputOutputLogger from "@middy/input-output-logger";
 import validator from "@middy/validator";
 import { Context } from "aws-lambda";
 import JSONErrorHandlerMiddleware from "middy-middleware-json-error-handler";
-import { CspRequest } from "../../models/cost-jobs";
-import { ProvisionRequest, provisionRequestSchema } from "../../models/provisioning-jobs";
+import {
+  AsyncProvisionRequest,
+  NewPortfolioPayload,
+  ProvisionCspResponse,
+  ProvisionRequest,
+  provisionRequestSchema,
+} from "../../models/provisioning-jobs";
 import { logger } from "../../utils/logging";
 import { errorHandlingMiddleware } from "../../utils/middleware/error-handling-middleware";
 import { ValidationErrorResponse } from "../../utils/response";
 import { tracer } from "../../utils/tracing";
-import { cspRequest, CspResponse } from "../util/csp-request";
+import { CspResponse } from "../util/csp-request";
+import { AtatClient, AtatApiError, IAtatClient } from "../client/client";
+import * as atatApiTypes from "../client/types";
+import { makeClient } from "../../utils/atat-client";
+
+function transformSynchronousResponse(
+  response: atatApiTypes.AddPortfolioResponseSync,
+  request: atatApiTypes.AddPortfolioRequest
+): CspResponse<atatApiTypes.AddPortfolioRequest, atatApiTypes.AddPortfolioResponseSync> {
+  return {
+    code: response.$metadata.status,
+    content: {
+      response,
+      request,
+    },
+  };
+}
+
+function transformAsynchronousResponse(
+  response: atatApiTypes.AddPortfolioResponseAsync,
+  request: atatApiTypes.AddPortfolioRequest
+): CspResponse<atatApiTypes.AddPortfolioRequest, atatApiTypes.AddPortfolioResponseAsync | { details: string }> {
+  if (response.location) {
+    return {
+      code: response.$metadata.status,
+      content: {
+        response,
+        request,
+      },
+    };
+  }
+  return {
+    code: 500,
+    content: {
+      response: {
+        details: "Location header was invalid or not provided",
+      },
+      request,
+    },
+  };
+}
+
+async function makeRequest(client: IAtatClient, request: ProvisionRequest): Promise<ProvisionCspResponse> {
+  // This function will always be operating for creating new portfolios; if we have something
+  // else, this will be a non-recoverable error anyway.
+  const payload = request.payload as NewPortfolioPayload;
```

Comment:
Should there be some guardrails to ensure a payload is properly created?
Something like:
```
if request.payload != null && request.payload != undefined

Reply:
I think middy captures that and it would get caught by a validation error for state input.